pkg/services/ngalert/notifier/receivers.go

```diff
@@ -0,0 +1,197 @@
+package notifier
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"time"
+
+	apimodels "github.com/grafana/grafana/pkg/services/ngalert/api/tooling/definitions"
+	"github.com/prometheus/alertmanager/notify"
+	"github.com/prometheus/alertmanager/types"
+	"github.com/prometheus/common/model"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	maxTestReceiversWorkers = 10
+)
+
+type InvalidReceiverError struct {
+	Receiver *apimodels.PostableGrafanaReceiver
+	Err      error
+}
+
+func (e InvalidReceiverError) Error() string {
+	return fmt.Sprintf("the receiver %s is invalid: %s", e.Receiver.Name, e.Err)
+}
+
+type ReceiverTimeoutError struct {
+	Receiver *apimodels.PostableGrafanaReceiver
+	Err      error
+}
+
+func (e ReceiverTimeoutError) Error() string {
+	return fmt.Sprintf("the receiver timed out: %s", e.Err)
+}
+
+func (am *Alertmanager) TestReceivers(ctx context.Context, c apimodels.TestReceiversConfig) (*apimodels.TestReceiversResult, error) {
+	// now represents the start time of the test
+	now := time.Now()
+	testAlert := &types.Alert{
+		Alert: model.Alert{
+			Labels: model.LabelSet{
+				model.LabelName("alertname"): "AlwaysFiring",
```

Comment:
I would suggest changing to something indicating that is actually a test (not actual) alert.